gapic/src/main/com/google/gapid/views/TracerDialog.java

```diff
@@ -498,20 +501,34 @@ protected void configureDialog(DirectoryDialog dialog) {
         requiredFieldMessage.setForeground(getDisplay().getSystemColor(SWT.COLOR_RED));
         requiredFieldMessage.setVisible(false);
 
-        Link adbWarning = withLayoutData(
-            createLink(this, "Path to adb invalid/missing. " +
-                "To trace on Android, please fix it in the <a>preferences</a>.",
-                e -> SettingsDialog.showSettingsDialog(getShell(), models, widgets.theme)),
+        if (!models.settings.isAdbValid()) {
+          Link adbWarning = withLayoutData(
+              createLink(this, "Path to adb invalid/missing. " +
+                  "To trace on Android, please fix it in the <a>preferences</a>.",
+                  e -> SettingsDialog.showSettingsDialog(getShell(), models, widgets.theme)),
+              new GridData(SWT.FILL, SWT.FILL, true, false));
+          adbWarning.setForeground(getDisplay().getSystemColor(SWT.COLOR_DARK_RED));
+        }
+
+        emptyAppWarning = withLayoutData(
+            createLabel(this, EMPTY_APP_WITH_RENDER_STAGE),
             new GridData(SWT.FILL, SWT.FILL, true, false));
-        adbWarning.setForeground(getDisplay().getSystemColor(SWT.COLOR_DARK_RED));
-        adbWarning.setVisible(!models.settings.isAdbValid());
+        emptyAppWarning.setForeground(getDisplay().getSystemColor(SWT.COLOR_DARK_YELLOW));
+        emptyAppWarning.setVisible(false);
+        device.getCombo().addListener(SWT.Selection, e -> updateEmptyAppWithRenderStageWarning(models.settings));
+        api.getCombo().addListener(SWT.Selection, e -> updateEmptyAppWithRenderStageWarning(models.settings));
+        traceTarget.addBoxListener(SWT.Modify, e -> updateEmptyAppWithRenderStageWarning(models.settings));
 
-        newAngleAvailableMessage = createLink(this, "A new ANGLE version is available for this " +
-            "device. Click to <a>download and install</a> the APK.",
+        RowLayout angleBarLayout = new RowLayout();
```

Comment:
Maybe the text message would be better placed above the button instead of next to it? I think this is a bit more common (I haven't found a good reference/guideline for this kind of UI design yet, just based on my intuition and a quick Google image search).

If it is next to it, it should probably also aligned in the middle vertically.

Reply:
We have the convention of allocating space for every hidden UI element, which is why there is already a lot of typically empty vertical space above the dialog button bar. This looks a bit weird already, and I don't want to keep it at a minimum, so I prefer the horizontal placemant. I will try to align it in the middle.

Reply:
The alignment is fixed. Thanks for paying attention to the details!